### PR TITLE
Speed up extraction when number of files is very large.

### PR DESCRIPTION
**Pull request by vladfi1**
On my system, most of the time was spent on basic python operations like list searches. Changing these to dictionary lookups sped things up significantly when extracting from an archive with ~100K files.

### CODE REVIEW

**Comment by vladfi1** [comment]
There's a mypy error specifically for python3.9, but I can't see what it is. Might have something to do with `targets` starting as a list but then being converted to a set?

**Comment by miurahr** [comment]
> py7zr/py7zr.py:558:23: error: Incompatible types in assignment (expression has type "set[str]", variable has type "Optional[list[str]]")  [assignment]

**Comment by miurahr** [comment]
You can find `read` and `extract` API accept `targets` as `List` and call `_extract` method and argument has `targets` as List

https://github.com/miurahr/py7zr/blob/8d0bcf263b52923793eb3e326d094481634caae5/py7zr/py7zr.py#L982C1-L987C56
```
    def read(self, targets: Optional[List[str]] = None) -> Optional[Dict[str, IO[Any]]]:
        self._dict = {}
        return self._extract(path=None, targets=targets, return_dict=True)

    def extract(self, path: Optional[Any] = None, targets: Optional[List[str]] = None) -> None:
        self._extract(path, targets, return_dict=False)
```

You can change these methods to accept both List and Set, and pass to _extract as set with conversion.

**Comment by vladfi1** [comment]
I've updated those methods to take a `Collection` instead.

**Review by miurahr — APPROVED** [review]

**Comment by miurahr** [comment]
Please update a coding style.
```
  check: commands[0]> flake8 py7zr tests
  py7zr/py7zr.py:40:1: I001 isort found an import in the wrong position
  py7zr/py7zr.py:41:1: I005 isort found an unexpected missing import
  tests/test_badfiles.py:91:10: BLK100 Black would make changes.
```

File: py7zr/py7zr.py [hunk 1]
```diff
@@ -37,7 +37,7 @@
 import sys
 from multiprocessing import Process
 from threading import Thread
-from typing import IO, Any, BinaryIO, Dict, List, Optional, Tuple, Type, Union
+from typing import IO, Any, BinaryIO, Dict, Collection, List, Optional, Tuple, Type, Union
```
Comment (miurahr) [inline py7zr/py7zr.py:40]:
```suggestion
from typing import IO, Any, BinaryIO, Collection, Dict, List, Optional, Tuple, Type, Union
```
sort order is wrong.

**Review by miurahr — CHANGES_REQUESTED** [review]

File: py7zr/py7zr.py (hunk 1 above)
Comment (miurahr) [reply py7zr/py7zr.py:40]: You can check it by `tox check`

**Comment by miurahr** [comment]
Please apply a patch below to avoid flake8-black error in unrelated test code.

```
From 0162869d6d19b248f38b04448156fc93b8543b25 Mon Sep 17 00:00:00 2001
From: Hiroshi Miura <miurahr@linux.com>
Date: Sat, 27 Jan 2024 11:56:10 +0900
Subject: [PATCH] chore: tweak flake8-black/black incompatibility

Signed-off-by: Hiroshi Miura <miurahr@linux.com>
---
 tests/test_badfiles.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tests/test_badfiles.py b/tests/test_badfiles.py
index 32e1e1e..a35b72a 100644
--- a/tests/test_badfiles.py
+++ b/tests/test_badfiles.py
@@ -90,7 +90,7 @@ def test_write_compressed_archive(tmp_path):
     contents = (Contents(filename="bin/qmake", text="qqqqq"), Contents(filename="lib/libhoge.so", text="hoge"))
     with TemporaryDirectory() as temp_path, SevenZipFile(
         tmp_path / "tools_qtcreator-linux-qt.tools.qtcreator.7z", "w"
-    ) as archive:
+    ) as archive:  # fmt: skip
         dest = pathlib.Path(temp_path)
         for folder in ("bin", "lib", "mkspecs"):
             (dest / folder).mkdir(parents=True, exist_ok=True)
@@ -102,5 +102,5 @@ def test_write_compressed_archive(tmp_path):
         archive.writeall(path=temp_path, arcname="target")
     with TemporaryDirectory() as target_path, SevenZipFile(
         tmp_path / "tools_qtcreator-linux-qt.tools.qtcreator.7z", "r"
-    ) as archive:
+    ) as archive:  # fmt: skip
         archive.extractall(path=target_path)
-- 
2.34.1

```

**Comment by vladfi1** [comment]
I've rebased master to pull in the test fix.

**Review by miurahr — APPROVED** [review]

**Comment by miurahr** [comment]
Thank you for a  hard work.

**Pull request merged**